Author: Subhash Kumar Katnapally <quic_skatnapa@quicinc.com>
Date:   Tue Feb 21 17:15:35 2023 +0530

    config-netifd: Avoid deleting MLO bond netdevice

    Avoid deleting MLO bond netdevice by netifd

    Change-Id: I1ccbe32c214a323d1d9014df041f77a8b1887e54
    Signed-off-by: Subhash Kumar Katnapally <quic_skatnapa@quicinc.com>

Index: netifd-2021-01-09-753c351b/system-linux.c
===================================================================
--- netifd-2021-01-09-753c351b.orig/system-linux.c
+++ netifd-2021-01-09-753c351b/system-linux.c
@@ -1141,12 +1141,43 @@ out:
 }

 /*
+ * system_get_bonding_mode()
+ *      Get bond mode
+ */
+static char *system_get_bonding_mode(const char *name, char *buf, const size_t buf_sz)
+{
+	int fd = -1;
+	char *path;
+
+	snprintf(buf, buf_sz, "/sys/class/net/%s/bonding/mode", name);
+	fd = open(buf, O_RDONLY);
+	if (fd < 0)
+		return NULL;
+
+	ssize_t len = read(fd, buf, buf_sz - 1);
+	if (len < 0) {
+		close(fd);
+		return NULL;
+	}
+
+	buf[len] = 0;
+	path = strrchr(buf, '/');
+	if (!path) {
+		close(fd);
+		return NULL;
+	}
+
+	close(fd);
+	return path + 1;
+}
+
+/*
  * Clear bridge (membership) state and bring down device
  */
 void system_if_clear_state(struct device *dev)
 {
 	static char buf[256];
-	char *bridge;
+	char *bridge, *mode;
 	struct device *bonding;

 	device_set_ifindex(dev, system_if_resolve(dev));
@@ -1155,10 +1186,19 @@ void system_if_clear_state(struct device
 		return;

 	if (system_is_bonding(dev->ifname, buf, sizeof(buf))) {
-		D(SYSTEM, "Delete existing bonding named '%s'\n", dev->ifname);
-		system_bonding_delbonding(dev);
+
+		/*
+		 * Do not delete bond netdevice with MLO type.
+		*/
+		mode = system_get_bonding_mode(dev->ifname, buf, sizeof(buf));
+		if (strstr(mode, "mlo") == NULL) {
+			D(SYSTEM, "Delete existing bonding named '%s'\n", dev->ifname);
+			system_bonding_delbonding(dev);
+		} else {
+			D(SYSTEM, "Do not delete mlo bonding named '%s'\n", dev->ifname);
+		}
 		return;
-	}
+        }

 	system_if_flags(dev->ifname, 0, IFF_UP);

